lib/modules/showImages.js

```diff
@@ -644,286 +636,366 @@ function domainToModuleName(hostname) {
 	return domainStack.pop();
 }
 
-async function checkElementForImage(elem, selfTextExpando = false) {
-	if (module.options.hideNSFW.value) {
-		elem.NSFW = new Thing(elem).isNSFW();
-	} else {
-		elem.NSFW = false;
+async function convertGifToVideo(options) {
+	try {
+		const info = await ajax({
+			type: 'json',
+			url: '//upload.gfycat.com/transcodeRelease',
+			data: { fetchUrl: options.src },
+			cacheFor: DAY,
+		});
+
+		if (!info.gfyName) throw new Error();
+
+		return {
+			options: await gfycat.handleLink('', [], info),
+			siteModule: gfycat,
+		};
+	} catch (e) {
+		throw Error('Could not convert gif to video', options.src);
 	}
+}
+
+async function checkElementForMedia(mediaLink) {
+	if (mediaLink.classList.contains('imgScanned')) return;
+	mediaLink.classList.add('imgScanned');
+
+	let thing;
 
-	elem.selfTextExpando = selfTextExpando;
-	elem.linklistExpando = !!elem.parentNode.classList.contains('title');
+	if (module.options.hideNSFW.value) {
+		thing = new Thing(mediaLink);
+		if (thing.isNSFW()) return;
+	}
 
-	const href = elem.href;
-	if ((!elem.classList.contains('imgScanned') && (typeof imagesRevealed[href] === 'undefined' || !module.options.ignoreDuplicates.value || isCurrentSubreddit('dashboard')) && href !== null)) {
-		elem.classList.add('imgScanned');
-		dupeAnchors++;
+	let detectResult;
+	let siteModule = [
+		'default',
+		'defaultVideo',
+		'defaultAudio',
+		domainToModuleName(mediaLink.hostname),
+	]
+		.map(modId => domainModuleMap[modId])
+		.find(mod => mod && (detectResult = mod.detect(mediaLink)));
+
+	if (!detectResult || !siteModule) return;
+
+	thing = thing || new Thing(mediaLink);
+	const inText = !!$(mediaLink).closest('.md').length;
+
+	const expando = new Expando(mediaLink, thing, inText);
+	expandos.set(mediaLink, expando);
+
+	let isDuplicate;
+	let reject, resolve, promise;
+	if (expandoPromises[mediaLink.href]) {
+		isDuplicate = true;
+		promise = expandoPromises[mediaLink.href];
+	} else {
+		expandoPromises[mediaLink.href] = promise =
+			new Promise((res, rej) => { resolve = res; reject = rej; });
+	}
 
-		let detectResult;
+	if (!inText) {
+		const originalExpando = thing.getExpandoButton();
+		if (originalExpando) {
+			if (thing.querySelector('.media-preview')) {
+				console.log('A media preview is already available; skipping.');
+				return;
+			}
 
-		const siteModule = [
-			'default',
-			'defaultVideo',
-			'defaultAudio',
-			domainToModuleName(elem.hostname),
-		]
-			.map(modId => domainModuleMap[modId])
-			.find(mod => mod && (detectResult = mod.detect(elem)));
+			originalExpando.remove();
+			promise.catch(() => {
+				expando.$button.after(originalExpando);
+			});
+		}
+	}
 
-		if (!siteModule) return;
+	expando.$button.insertAfter(!inText && thing.getTitleElement() ?
+		mediaLink.parentElement : mediaLink);
 
-		elem.site = siteModule.moduleID;
+	try {
+		if (module.options.ignoreDuplicates.value && isDuplicate) {
+			expando.buildDuplicate(await promise);
```

Comment:
I made an attempt at cleaning it up now. I also removed the `ignoreDuplicates` option, since the UX issues associated with the previous solution are gone.